### PR TITLE
Replace System.Drawing.Bitmap with Png for image saving

### PR DESCRIPTION
**Pull request by pchote**
This PR continues the work from #15973 by extending our `Png` class to support image saving (using `Bitmap` internally for now, to be replaced in a followup PR), and porting all consumers to the new API.

This also changes the way screenshot saving works so that more work can be done from the worker thread.  This should reduce the size of the perf hitch when saving screenshots.

Dependency for #15973.

### CODE REVIEW

**Review by reaperrr — DISMISSED** [review]
The code is out of my league, but screenshots and map preview saving work fine and no other regressions spotted either, so 👍

File: OpenRA.Game/Graphics/Sheet.cs [hunk 1]
```diff
@@ -11,9 +11,8 @@
 
 using System;
 using System.Drawing;
-using System.Drawing.Imaging;
 using System.IO;
-using System.Runtime.InteropServices;
+using System.Linq;
```
Comment (teinarss) [inline OpenRA.Game/Graphics/Sheet.cs]: Not needed

File: OpenRA.Platforms.Default/ThreadedGraphicsContext.cs [hunk 2]
```diff
@@ -14,6 +14,7 @@
 using System.Drawing;
 using System.Runtime.ExceptionServices;
 using System.Threading;
+using OpenRA.FileFormats;
```
Comment (teinarss) [inline OpenRA.Platforms.Default/ThreadedGraphicsContext.cs]: Also not needed :)

File: OpenRA.Game/FileFormats/Png.cs [hunk 3]
```diff
@@ -206,5 +228,116 @@ static bool IsPaletted(byte bitDepth, PngColorType colorType)
...
+
+					bitmap.Save(bitmapStream, ImageFormat.Png);
+				}
+
+				if (!EmbeddedData.Any())
+					return bitmapStream.ToArray();
+
+				// Add embedded metadata to the end of the image
+				bitmapStream.Position = 0;
+				var outputStream = new MemoryStream();
```
Comment (teinarss) [inline OpenRA.Game/FileFormats/Png.cs]: Missing using()

**Review by teinarss — CHANGES_REQUESTED** [review]

File: OpenRA.Game/FileFormats/Png.cs (hunk 3 above)
Comment (pchote) [reply OpenRA.Game/FileFormats/Png.cs]: The `BinaryWriter` disposes the stream at the end of its using - one of the reasons I don't like about it, but I didn't see a strong reason to rewrite @IceReaper's initial use of it as this whole method needs to be replaced in the short term to remove `System.Bitmap`.

**Comment by pchote** [comment]
Rebased and updated.

File: OpenRA.Platforms.Default/Sdl2GraphicsContext.cs [hunk 4]
```diff
@@ -111,39 +115,51 @@ public void DisableScissor()
...
 
-			OpenGL.glPushClientAttrib(OpenGL.GL_CLIENT_PIXEL_STORE_BIT);
-
-			OpenGL.glPixelStoref(OpenGL.GL_PACK_ROW_LENGTH, data.Stride / 4f);
-			OpenGL.glPixelStoref(OpenGL.GL_PACK_ALIGNMENT, 1);
+			unsafe
+			{
+				fixed (byte* pRaw = raw)
+				{
+					OpenGL.glPushClientAttrib(OpenGL.GL_CLIENT_PIXEL_STORE_BIT);
```
Comment (chrisforbes) [inline OpenRA.Platforms.Default/Sdl2GraphicsContext.cs]: I would prefer to keep the unsafe/pinned block as small as possible. This state mangling doesn't need to be inside.

**Review by chrisforbes — DISMISSED** [review]

**Comment by pchote** [comment]
Fixed and rebased again.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ba283c4d93d23996858869627f5b892da5864299...861e39bd866e1832b89a8f65a9b53fba4fb00cdf

**Comment by IceReaper** [comment]
Ah already merged :) Anyway would have gotten an👍 from me

**Pull request merged**